samples/nrf9160/nrf_cloud_rest_device_message/src/main.c

```diff
@@ -0,0 +1,429 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <modem/nrf_modem_lib.h>
+#include <nrf_modem_at.h>
+#include <modem/modem_info.h>
+#include <settings/settings.h>
+#include <net/nrf_cloud.h>
+#include <net/nrf_cloud_rest.h>
+#include <sys/reboot.h>
+#include <logging/log.h>
+#include <dk_buttons_and_leds.h>
+
+LOG_MODULE_REGISTER(nrf_cloud_rest_device_message,
+				    CONFIG_NRF_CLOUD_REST_DEVICE_MESSAGE_SAMPLE_LOG_LEVEL);
+
+#define BTN_NUM			CONFIG_REST_DEVICE_MESSAGE_BUTTON_EVT_NUM
+#define LED_NUM			CONFIG_REST_DEVICE_MESSAGE_LED_NUM
+#define JITP_REQ_WAIT_SEC	10
+
+/* Semaphore to indicate a button has been pressed */
+static K_SEM_DEFINE(button_press_sem, 0, 1);
+
+/* Semaphore to indicate that a network connection has been established */
+static K_SEM_DEFINE(lte_connected, 0, 1);
+
+/* nRF Cloud device ID */
+static char device_id[NRF_CLOUD_CLIENT_ID_MAX_LEN + 1];
+
+#define REST_RX_BUF_SZ 2100
+/* Buffer used for REST calls */
+static char rx_buf[REST_RX_BUF_SZ];
+
+#define JWT_DURATION_S	(60*5)
+#define JWT_BUF_SZ	900
+/* Buffer used for JSON Web Tokens (JWTs) */
+static char jwt[JWT_BUF_SZ];
+
+/* nRF Cloud REST context */
+struct nrf_cloud_rest_context rest_ctx = {
+	.connect_socket = -1,
+	.keep_alive = true,
+	.rx_buf = rx_buf,
+	.rx_buf_len = sizeof(rx_buf),
+	.fragment_size = 0
+};
+
+/* Flag to indicate if the user requested JITP to be performed */
+static bool jitp_requested;
+
+int set_led(const int state)
+{
+	int err = dk_set_led(LED_NUM, state);
+
+	if (err) {
+		LOG_ERR("Failed to set LED, error: %d", err);
+		return err;
+	}
+	return 0;
+}
+
+int init_led(void)
+{
+	int err = dk_leds_init();
+
+	if (err) {
+		LOG_ERR("LED init failed, error: %d", err);
+		return err;
+	}
+
+	return set_led(0);
+}
+
+static void button_handler(uint32_t button_states, uint32_t has_changed)
+{
+	if (has_changed & button_states & BIT(BTN_NUM - 1)) {
+		LOG_DBG("Button %d pressed", BTN_NUM);
+		k_sem_give(&button_press_sem);
+	}
+}
+
+static int generate_jwt(void)
+{
+	int err = nrf_cloud_jwt_generate(JWT_DURATION_S, jwt, sizeof(jwt));
+
+	if (err < 0) {
+		LOG_ERR("Failed to generate JWT, error: %d", err);
+		return err;
+	}
+
+	LOG_DBG("JWT:\n%s", log_strdup(jwt));
+	rest_ctx.auth = jwt;
+
+	return 0;
+}
+
+int send_message_on_button(void)
+{
+	/* Wait for a buttonpress */
+	(void)k_sem_take(&button_press_sem, K_FOREVER);
+
+	/* Generate a new JWT */
+	int err = generate_jwt();
+
+	if (err) {
+		LOG_ERR("Failed to generate JWT for action %d", err);
+		return err;
+	}
+
+	/* Send off a JSON message about the buttonpress */
+	err = nrf_cloud_rest_send_device_message(&rest_ctx, device_id,
+			"{\"example_event\":\"button_press\"}", false, NULL);
```

Comment:
You got it!

I am opting not to import nrf_cloude_codec.h and use the defines there for a couple of reasons:
1. We are using a non-standard schema for the hello world message, so total conformity is not a goal
2. I think it would detract from the readability of the sample

Reply:
use the schema and just send a press